src/probe/new_program_iterator.jl

```diff
@@ -0,0 +1,69 @@
+struct NewProgramsIterator
+    level::Int
+    bank::Vector{Vector{RuleNode}}
+    grammar::ContextSensitiveGrammar
+end
+
+mutable struct NewProgramsState
+    rule_index::Int
+    sum_iter::Union{SumIterator,Nothing}
+    sum_iter_state::Union{Tuple{Vector{Int},SumIteratorState},Nothing}
+    cartesian_iter
+    cartesian_iter_state
+    index_program_option::Int
+end
+
+function Base.iterate(iter::NewProgramsIterator)
+    iterate(iter, NewProgramsState(1, SumIterator(0, 0, 0), nothing, Iterators.product(1:1), nothing, 1))
```

Comment:
I think that `sum_iter` should be set to `nothing` at first, otherwise it won't generate programs with the first rule if it is nonterminal.
```suggestion
    iterate(iter, NewProgramsState(1, nothing, nothing, Iterators.product(1:1), nothing, 1))
```

Reply:
Yes, you are right. This is indeed a bug